src/coreclr/jit/morph.cpp

```diff
@@ -5303,8 +5303,9 @@ GenTree* Compiler::fgMorphField(GenTree* tree, MorphAddrContext* mac)
             // then pFldAddr will be NULL and
             //      fldAddr will be the actual address of the static field
             //
-            void** pFldAddr = nullptr;
-            void*  fldAddr  = info.compCompHnd->getFieldAddress(symHnd, (void**)&pFldAddr);
+            void*  boxedStatic = nullptr;
+            void** pFldAddr    = nullptr;
+            void*  fldAddr     = info.compCompHnd->getFieldAddress(symHnd, (void**)&pFldAddr, &boxedStatic);
```

Comment:
This design where how we access the field can effectively change between importer and morph is problematic for field sequence attachment.

For boxes statics, the field sequence will record an offset of `4/8` (the method table pointer offset).

For statics with a known address, it will record that address.

The implication is that if the given field "flips" from being a boxed static to a static with a known address, the field sequences we obtained for it become wrong, and to VN (and other possible consumers of field sequences) it will look at though all access to that field is out-of-bounds.

It is very unfortunate that the way these statics are to be accesses can effectively change during the compilation. It is not clear to me how would the proper fix look like (for `ldsfld` we can carry the `FIELD<struct>` node into morph, but for `ldsflda` we currently don't create any `FIELD` nodes at all).

Reply:
Yes, I also had asserts with changed fields sequences. My understanding that we'd better limit this optimization to initialized types only, by the time we reach tier1 we should initialize all hot types anyway, right?

Reply:
Presumably.

I don't have a good handle on the tradeoffs here, but it would be great to avoid presenting this "shifting" view to the Jit.

Reply:
Reverted the previous version and limited to `IsClassInited`